File: stack & queue/CircularQueueArray.java

public class CircularQueueArray {
    int[] array;
    int front, rear;
    int capacity;

    CircularQueueArray(int capacity){
        this.capacity = capacity;
        front = rear = -1;
        array = new int[capacity];
    }

    boolean isFull(){
        return front == rear+1 || front == 0 && rear == capacity-1;
    }

    boolean isEmpty(){
        return front == -1;
    }

    void enqueue(int key){
        if (isFull()){
            System.out.println("queue is full");
            return;
        }
        if (front == -1)
            front = 0;
        rear = (rear+1)%capacity;
        array[rear] = key;
    }

    int dequeue(){
        if (isEmpty())
            return -1;
        if (front>=rear)
            front = rear = -1;
        front = (front+1)%capacity;
        return array[front];
    }

    int front(){
        return array[front];
    }

    int rear(){
        return array[rear];
    }
    
    void print(){
        int i;
        for (i=front; i!=rear; i=(i+1)%capacity){
            System.out.print(array[i]+" ");
        }
        System.out.print(array[i]+" ");
    }

    public static void main(String[] args) {
        
    }
}

File: queue/CircularQueueArray.js

class CircularQueueArray {
    constructor(capacity) {
        this.size = capacity;
        this.front = -1;
        this.rear = -1;
        this.items = [];
    }

    isFull() {
        return this.front === ((this.rear+1)%this.size);
    }

    isEmpty() {
        return this.front === -1 && this.rear === -1;
    }

    peek() {
        return this.items[this.front];
    }

    enqueue(key) {
        if (this.isFull()) {
            console.log('queue is full');
            return;
        }
        if (this.front === -1) {
            this.front = 0;
        }
        this.rear = (this.rear+1)%this.size
        this.items[this.rear] = key;
    }

    dequeue() {
        if (this.isEmpty()) {
            console.log('queue is empty');
            return;
        }
        let key = this.items[this.front];
        if (this.front === this.rear) {
            this.front = this.rear = -1;
        } else {
            this.front = ((this.front+1)%this.size);
        }
        return key;
    }
}

export default CircularQueueArray;

/*  Usage
    const queue = new CircularQueueArray(4);
    queue.enqueue(3);
    queue.enqueue(4);
    queue.enqueue(5);
    queue.enqueue(6);
    console.log(queue.peek())
    console.log(queue.dequeue())
    console.log(queue.peek())
*/